include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -1160,6 +1293,144 @@ namespace VectorTools
+    typename MatrixFree<dim,number>::AdditionalData additional_data;
+    additional_data.tasks_parallel_scheme =
+      MatrixFree<dim,number>::AdditionalData::partition_color;
+    if (parallel_tria)
+      {
+        const MPI_Comm &mpi_communicator = parallel_tria->get_communicator();
+        additional_data.mpi_communicator = mpi_communicator;
+      }
+    additional_data.mapping_update_flags = (update_values | update_JxW_values);
+    MatrixFree<dim, number> matrix_free;
+    matrix_free.reinit (mapping, dof, constraints,
+                        QGauss<1>(fe_degree+2), additional_data);
+    typedef MatrixFreeOperators::MassOperator<dim, fe_degree, components, number> MatrixType;
+    MatrixType mass_matrix;
+    mass_matrix.initialize(matrix_free);
+    mass_matrix.compute_diagonal();
+
+    typedef LinearAlgebra::distributed::Vector<number> LocalVectorType;
+    LocalVectorType vec, rhs, inhomogeneities;
+    matrix_free.initialize_dof_vector(vec);
+    matrix_free.initialize_dof_vector(rhs);
+    matrix_free.initialize_dof_vector(inhomogeneities);
+    constraints.distribute(inhomogeneities);
+    inhomogeneities*=-1.;
+
+    create_right_hand_side (mapping, dof, quadrature, function, rhs, constraints);
+    mass_matrix.vmult_add(rhs, inhomogeneities);
+
+    //now invert the matrix
+    ReductionControl     control(5*rhs.size(), 0., 1e-12, false, false);
+    SolverCG<LinearAlgebra::distributed::Vector<number> > cg(control);
+    PreconditionJacobi<MatrixType> preconditioner;
+    preconditioner.initialize(mass_matrix, 1.);
+    cg.solve (mass_matrix, vec, rhs, preconditioner);
+    vec+=inhomogeneities;
+
+    constraints.distribute (vec);
+    IndexSet::ElementIterator it = locally_owned_dofs.begin();
+    for (; it!=locally_owned_dofs.end(); ++it)
+      vec_result(*it) = vec(*it);
+    vec_result.compress(VectorOperation::insert);
+  }
+
+
+
+  template <int dim, typename VectorType>
+  void project_matrix_free (const Mapping<dim>   &mapping,
+                            const DoFHandler<dim> &dof,
+                            const ConstraintMatrix         &constraints,
+                            const Quadrature<dim>          &quadrature,
+                            const Function<dim, typename VectorType::value_type> &function,
+                            VectorType                     &vec_result,
+                            const bool                     enforce_zero_boundary,
+                            const Quadrature<dim-1>        &q_boundary,
+                            const bool                     project_to_boundary_first)
+  {
+    // If we can, use the matrix-free implementation
+    bool use_matrix_free = true;
+    // enforce_zero_boundary and project_to_boundary_first
+    // are not yet supported
+    if (enforce_zero_boundary || project_to_boundary_first)
+      use_matrix_free = false;
+    else
+      {
+        // Find out if the FiniteElement is supported
```

Comment:
The check is currently done in `include/deal.II/matrix_free/shape_info.templates.h`, lines 79-113. I think it would indeed be a good idea to have a check in `MatrixFree` which could just use some logic in `shape_info.h`. I can implement that next week since I'm busy both today and tomorrow. If somebody beats me until then, I'm happy too.